File: packages/lynx/host/ios/OpenChamberLynxOAuthBrowser.swift
import AuthenticationServices
import Foundation

/// ASWebAuthenticationSession launcher stub.
/// Mirrors `packages/lynx/src/host/oauthBrowser.ts`.
/// Inject point: present from the host window's presentationAnchor.
enum OpenChamberLynxOAuthBrowserResult {
  case ok(callbackUrl: URL?)
  case cancelled
  case unavailable
  case failed(String)
}

protocol OpenChamberLynxOAuthBrowsing: AnyObject {
  func openAuthorize(
    url: URL,
    callbackScheme: String,
    prefersEphemeral: Bool
  ) async -> OpenChamberLynxOAuthBrowserResult
}

/// Stub session factory. Real device binary must set presentationContextProvider.
final class OpenChamberLynxOAuthBrowser: NSObject, OpenChamberLynxOAuthBrowsing, ASWebAuthenticationPresentationContextProviding {
  weak var presentationAnchor: ASPresentationAnchor?

  func openAuthorize(
    url: URL,
    callbackScheme: String,
    prefersEphemeral: Bool
  ) async -> OpenChamberLynxOAuthBrowserResult {
    // Without a presentation anchor (Linux/scaffold) stay unavailable — never fake OAuth.
    guard presentationAnchor != nil else { return .unavailable }
    return await withCheckedContinuation { continuation in
      let session = ASWebAuthenticationSession(
        url: url,
        callbackURLScheme: callbackScheme
      ) { callbackURL, error in
        if let error = error as? ASWebAuthenticationSessionError,
           error.code == .canceledLogin {
          continuation.resume(returning: .cancelled)
          return
        }
        if let error {
          continuation.resume(returning: .failed(error.localizedDescription))
          return
        }
        continuation.resume(returning: .ok(callbackUrl: callbackURL))
      }
      session.presentationContextProvider = self
      session.prefersEphemeralWebBrowserSession = prefersEphemeral
      if !session.start() {
        continuation.resume(returning: .failed("ASWebAuthenticationSession failed to start"))
      }
    }
  }

  func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
    presentationAnchor ?? ASPresentationAnchor()
  }
}

File: packages/lynx/host/android/app/src/main/java/com/yee94/openchamber/lynx/OpenChamberLynxOAuthBrowser.kt
package com.yee94.openchamber.lynx

/**
 * Custom Tabs OAuth launcher stub.
 * Mirrors packages/lynx/src/host/oauthBrowser.ts
 * Inject: CustomTabsIntent + openchamber:// intent-filter resume.
 */
sealed class LynxOAuthBrowserResult {
    data class Ok(val callbackUrl: String?) : LynxOAuthBrowserResult()
    object Cancelled : LynxOAuthBrowserResult()
    object Unavailable : LynxOAuthBrowserResult()
    data class Failed(val error: String) : LynxOAuthBrowserResult()
}

interface LynxOAuthBrowsing {
    suspend fun openAuthorize(
        url: String,
        callbackScheme: String,
        prefersEphemeral: Boolean,
    ): LynxOAuthBrowserResult
}

class OpenChamberLynxOAuthBrowser(
    private val launcher: (suspend (url: String, callbackScheme: String) -> LynxOAuthBrowserResult)? = null,
) : LynxOAuthBrowsing {
    override suspend fun openAuthorize(
        url: String,
        callbackScheme: String,
        prefersEphemeral: Boolean,
    ): LynxOAuthBrowserResult {
        // Inject point:
        // CustomTabsIntent.Builder().build().launchUrl(activity, Uri.parse(url))
        // + Activity result / deep link for `$callbackScheme://...`
        if (launcher == null) return LynxOAuthBrowserResult.Unavailable
        @Suppress("UNUSED_VARIABLE")
        val ephemeral = prefersEphemeral
        return launcher.invoke(url, callbackScheme)
    }
}
